Hello there!

My name is Ernest and I will review your project. I hope you enjoyed working on it.  
If there are any inaccuracies in the solution, then I will draw your attention to them. Please try to find the error and fix it first.  
However, if there are difficulties, please, let me know. I will be able to clarify and suggest something more.

Below you will find my comments - **please do not move, modify or delete them**.

You can find my comments in green, yellow or red boxes like this:

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Success. Everything is done succesfully.
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Remarks. Some recommendations.
</div>

<div class="alert alert-block alert-danger">

<b>Reviewer's comment</b> <a class="tocSkip"></a>

Needs fixing. The block requires some corrections. Work can't be accepted with the red comments.
</div>

You can answer me by using this:

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>

Text here.
</div>

# Introduction

You work for the OilyGiant mining company. Your task is to find the best place for a new well.

You have data on oil samples from three regions. Parameters of each oil well in the region are already known. Build a model that will help to pick the region with the highest profit margin. Analyze potential profit and risks

## Initialization

### Load necessary libraries

In [1]:
# Load necessary libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

### Load up data

In [2]:
# First dataset
data_1 = pd.read_csv('/datasets/geo_data_0.csv')

In [3]:
# Second dataset
data_2 = pd.read_csv('/datasets/geo_data_1.csv')

In [4]:
# Third dataset
data_3 = pd.read_csv('/datasets/geo_data_2.csv')

All necessary libraries have been loaded and all the data is ready for preprocessing 

## Data Preprocessing

### Visualize datasets and view info

#### First dataset (data_1)

In [5]:
# Visualize first Dataset
data_1

id        f0        f1        f2     product
0      txEyH  0.705745 -0.497823  1.221170  105.280062
1      2acmU  1.334711 -0.340164  4.365080   73.037750
2      409Wp  1.022732  0.151990  1.419926   85.265647
3      iJLyR -0.032172  0.139033  2.978566  168.620776
4      Xdl7t  1.988431  0.155413  4.751769  154.036647
...      ...       ...       ...       ...         ...
99995  DLsed  0.971957  0.370953  6.075346  110.744026
99996  QKivN  1.392429 -0.382606  1.273912  122.346843
99997  3rnvd  1.029585  0.018787 -1.348308   64.375443
99998  7kl59  0.998163 -0.528582  1.583869   74.040764
99999  1CWhH  1.764754 -0.266417  5.722849  149.633246

[100000 rows x 5 columns]

In [6]:
# View data_1 info
data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


#### Second dataset (data_2)

In [7]:
# Visualize Second Dataset
data_2

id         f0         f1        f2     product
0      kBEdx -15.001348  -8.276000 -0.005876    3.179103
1      62mP7  14.272088  -3.475083  0.999183   26.953261
2      vyE1P   6.263187  -5.948386  5.001160  134.766305
3      KcrkZ -13.081196 -11.506057  4.999415  137.945408
4      AHL4O  12.702195  -8.147433  5.004363  134.766305
...      ...        ...        ...       ...         ...
99995  QywKC   9.535637  -6.878139  1.998296   53.906522
99996  ptvty -10.160631 -12.558096  5.005581  137.945408
99997  09gWa  -7.378891  -3.084104  4.998651  137.945408
99998  rqwUm   0.665714  -6.152593  1.000146   30.132364
99999  relB0  -3.426139  -7.794274 -0.003299    3.179103

[100000 rows x 5 columns]

In [8]:
# View data_2 info
data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


#### Third dataset (data_3)

In [9]:
# VIsualize Third Dataset
data_3

id        f0        f1        f2     product
0      fwXo0 -1.146987  0.963328 -0.828965   27.758673
1      WJtFt  0.262778  0.269839 -2.530187   56.069697
2      ovLUW  0.194587  0.289035 -5.586433   62.871910
3      q6cA6  2.236060 -0.553760  0.930038  114.572842
4      WPMUX -0.515993  1.716266  5.899011  149.600746
...      ...       ...       ...       ...         ...
99995  4GxBu -1.777037  1.125220  6.263374  172.327046
99996  YKFjq -1.261523 -0.894828  2.524545  138.748846
99997  tKPY3 -1.199934 -2.957637  5.219411  157.080080
99998  nmxp2 -2.419896  2.417221 -5.548444   51.795253
99999  V9kWn -2.551421 -2.025625  6.090891  102.775767

[100000 rows x 5 columns]

In [10]:
# View data_3 info 
data_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


At first glance, the dataset appears clean. There are no missing values and all columns have the correct datatype.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Wery well
</div>


### Check for duplicates

In [11]:
# Check for duplicates in data_1
print(data_1.duplicated().value_counts())

False    100000
dtype: int64


In [12]:
# Check for duplicates in data_2
print(data_2.duplicated().value_counts())

False    100000
dtype: int64


In [13]:
# Check for duplicates in data_3
print(data_3.duplicated().value_counts())

False    100000
dtype: int64


There are no duplicates in our datasets. The data is ready for manipulation

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

OK
</div>


## Model creation, training and evaluation

### Split data into training and validation sets

#### Split data_1

In [14]:
# Split data_1
d1_train, d1_valid = train_test_split(data_1, test_size=0.25, random_state=12345)

In [15]:
# Visualize d1_train
d1_train

id        f0        f1         f2     product
27212  sn8dh  0.022450  0.951034   2.197333  147.370612
7866   oqj58  1.766731  0.007835   6.436602  147.630053
62041  MaWjy  0.724514  0.666063   1.840177   77.696728
70185  jyV1V -1.104181  0.255268   2.026156   55.210501
82230  g2kgI -0.635263  0.747990   6.643327  113.891723
...      ...       ...       ...        ...         ...
4094   osNdz  1.863680 -0.298123   1.621324  124.380793
85412  YFN4o -1.162682 -0.014822   6.819941  144.874913
2177   Ay91j  0.862688 -0.403776   1.867662  134.967255
77285  MfaH6  0.846235 -0.489533   1.058786   64.494357
86498  qiCZa  2.019850  0.263887  11.497239  151.514894

[75000 rows x 5 columns]

In [16]:
# Visualize d1_valid
d1_valid

id        f0        f1        f2     product
71751  7Z2gx  0.948970 -0.057547  2.095727   10.038645
80493  b2HU8  0.992974  0.206671 -0.142278  114.551489
2655   Y0dXD  1.199854 -0.563356 -1.852991  132.603635
53233  Mje59  0.691422 -0.433347  0.564974  169.072125
91141  yqXxU  0.420772  0.972638  0.736190  122.325180
...      ...       ...       ...       ...         ...
12581  CzptN  1.841106 -0.107678  2.585648  170.116726
18456  cT9bJ  0.041859  0.489280  2.168001   93.632175
73035  Y7xR6  0.376389  1.054298 -0.430104  127.352259
63834  4WJRf  1.071330 -0.470046  4.527099   99.782700
43558  2RU3S  1.726318 -0.077871  5.007083  177.821022

[25000 rows x 5 columns]

#### Split data_2

In [17]:
# Split data_2
d2_train, d2_valid = train_test_split(data_2, test_size= 0.25, random_state= 12345)

In [18]:
# Visualize d2_train
d2_train

id         f0         f1        f2     product
27212  4jD5S  -6.488552  -1.590478  3.001311   84.038886
7866   jAjCc  18.819463   4.602079  2.996867   80.859783
62041  fPN1n  10.816499  -3.919653  1.991077   53.906522
70185  DhcUq -12.416362  -9.343774  0.996691   30.132364
82230  fhr6C -15.041012  -8.474624  1.996463   57.085625
...      ...        ...        ...       ...         ...
4094   Znkmi  -1.859774  -4.818321  2.998956   84.038886
85412  ONSGU   4.863108  -6.071357  3.990522  110.992147
2177   toesY  10.527491  -8.124357  5.004246  134.766305
77285  xiCBq   0.110801 -10.896529  3.005078   84.038886
86498  JWLV5  -4.653214  -4.283936  2.999043   84.038886

[75000 rows x 5 columns]

In [19]:
# Visualize d2_valid
d2_valid

id         f0         f1        f2     product
71751  aTbUt  -0.371866  -1.862494  3.002210   80.859783
80493  CK34V   9.015122 -13.881455  1.995363   53.906522
2655   ioXCO  -6.507568  -4.817448  1.003449   30.132364
53233  xk6tF  14.560845 -10.667755  1.995175   53.906522
91141  F9AmG   6.090476  -4.494723  0.013815    0.000000
...      ...        ...        ...       ...         ...
12581  fJBfd  -4.107455  -4.607004  4.991140  137.945408
18456  paaof  -8.363585  -6.357527  3.995592  110.992147
73035  tOADJ  -9.136241  -5.403258  5.000925  137.945408
63834  AiZ1G  -6.386442  -0.764568  3.011526   84.038886
43558  2PENr  10.542817  -6.206446  1.992290   53.906522

[25000 rows x 5 columns]

#### Split data_3

In [20]:
# Split data_3
d3_train, d3_valid = train_test_split(data_3, test_size= 0.25, random_state=12345)

In [21]:
# Visualize d3_train
d3_train

id        f0        f1        f2     product
27212  VtNl5 -0.906828  1.339033  1.106789   16.733577
7866   nUruK -1.536353 -0.704673 -1.747632   38.047492
62041  eipxz -1.959584  0.356044  3.528659  116.149369
70185  IGIba  2.129734  2.713459 -0.156172  105.888803
82230  Hvolv -0.332026  1.515604  5.417583   27.570897
...      ...       ...       ...       ...         ...
4094   P7iD3 -3.139524  0.684457  8.820754  155.009949
85412  OoGpc -0.471084 -2.195627  3.350739  143.348954
2177   BRM5L  0.642839  1.375212  4.809140   77.290797
77285  Yelrt -4.222809  0.191923  4.029111  121.243834
86498  gy5GR -2.996164  0.676341 -2.447329   39.377889

[75000 rows x 5 columns]

In [22]:
# Visualize d3_valid
d3_valid

id        f0        f1        f2     product
71751  DMlgl -1.444717 -3.861599  2.225805   61.212375
80493  tbrRC -1.418617  1.276544 -0.976433   41.850118
2655   JRs1w -4.587649 -0.413199  1.649268   57.776581
53233  bidOx  1.871584  1.619101  4.273555  100.053761
91141  AcBBB -2.028785  4.128167  6.089547  109.897122
...      ...       ...       ...       ...         ...
12581  y2n7n -1.315037 -0.122572 -0.345986   28.492402
18456  aBi2q  0.015144  0.876998  2.603666   21.431303
73035  CMYf3  0.725767 -0.151545  3.258620  125.487229
63834  HzkKW -0.232209 -0.180354 -0.524997   99.422903
43558  Qwd5N -2.350293 -1.041230  8.458378  127.445075

[25000 rows x 5 columns]

The datasets have been split into training and validation datsets in the ratio 75%:25%. Looking at the data columns, `id` doesn't seem to be informative when it comes to our task, so we will drop it in the next stage.

### Prepare features and target

#### d1_train

In [23]:
# Create features and target for d1_train
d1_features_train = d1_train.drop(['product', 'id'], axis=1)
d1_target_train = d1_train['product']

#### d1_valid

In [24]:
# Create features and target for d1_valid
d1_features_valid = d1_valid.drop(['product','id'], axis=1)
d1_target_valid = d1_valid['product']

#### d2_train

In [25]:
# Create features and target for d2_train
d2_features_train = d2_train.drop(['product', 'id'], axis=1)
d2_target_train = d2_train['product']

#### d2_valid

In [26]:
# Create features and target for d2_valid
d2_features_valid = d2_valid.drop(['product', 'id'], axis=1)
d2_target_valid = d2_valid['product']

#### d3_train

In [27]:
# Create features and target for d3_train
d3_features_train = d3_train.drop(['product', 'id'], axis=1)
d3_target_train = d3_train['product']

#### d3_valid

In [28]:
# Create features and target for d3_valid
d3_features_valid = d3_valid.drop(['product', 'id'], axis=1)
d3_target_valid = d3_valid['product']

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

It is more common to remove unnecessary like id before the split, but it is not critical.
</div>


### Standardize numeric columns

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

It is a good idea to check the distibution of features before scaling. Probably we will not need it at all.
</div>


#### d1_features_train

In [29]:
# Create and fit scaler_1
scaler_1 = StandardScaler()
scaler_1.fit(d1_features_train)

StandardScaler()

In [30]:
# Transform columns
d1_features_train = scaler_1.transform(d1_features_train)
d1_features_valid = scaler_1.transform(d1_features_valid)

#### d2_features_train

In [31]:
# Create and fit scaler_2
scaler_2 = StandardScaler()
scaler_2.fit(d2_features_train)

StandardScaler()

In [32]:
# Transform columns
d2_features_train = scaler_2.transform(d2_features_train)
d2_features_valid = scaler_2.transform(d2_features_valid)

#### d3_features_train

In [33]:
# Create and fit scaler_3
scaler_3 = StandardScaler()
scaler_3.fit(d3_features_train)

StandardScaler()

In [34]:
# Transform columns
d3_features_train = scaler_3.transform(d3_features_train)
d3_features_valid = scaler_3.transform(d3_features_valid)

All columns have been standardized

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

We should use transform method on validation features as well.
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment V2</b> <a class="tocSkip"></a>

OK
</div>


### Create and train models

#### Model_1

In [35]:
model_1 = LinearRegression()
model_1.fit(d1_features_train, d1_target_train)

LinearRegression()

#### Model_2

In [36]:
model_2 = LinearRegression()
model_2.fit(d2_features_train, d2_target_train)

LinearRegression()

#### Model_3

In [37]:
model_3 = LinearRegression()
model_3.fit(d3_features_train, d3_target_train)

LinearRegression()

Models are trained and ready for evaluation

### Model evaluation

#### Model_1

In [38]:
predict_1 = model_1.predict(d1_features_valid)

mse_1 = mean_squared_error(d1_target_valid, predict_1)

rmse_1 = mse_1 ** 0.5

ave_vol_1 = pd.Series(predict_1).mean()


print('RMSE:', rmse_1)
print('Average volume of predicted reserves:', ave_vol_1)

RMSE: 37.5794217150813
Average volume of predicted reserves: 92.59256778438035


#### Model_2

In [39]:
predict_2 = model_2.predict(d2_features_valid)

mse_2 = mean_squared_error(d2_target_valid, predict_2)

rmse_2 = mse_2 ** 0.5

ave_vol_2 = pd.Series(predict_2).mean()


print('RMSE:', rmse_2)
print('Average volume of predicted reserves:', ave_vol_2)

RMSE: 0.893099286775617
Average volume of predicted reserves: 68.728546895446


<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

You forgot to update the model and applied the wrong one here. That it why it is better to use functions, this way we do not need to update anything manually :)
</div>




<div class="alert alert-block alert-success">
<b>Reviewer's comment V2</b> <a class="tocSkip"></a>

OK
</div>


#### Model_3 

In [40]:
predict_3 = model_3.predict(d3_features_valid)

mse_3 = mean_squared_error(d3_target_valid, predict_3)

rmse_3 = mse_3 ** 0.5

ave_vol_3 = pd.Series(predict_3).mean()


print('RMSE:', rmse_3)
print('Average volume of predicted reserves:', ave_vol_3)

RMSE: 40.02970873393434
Average volume of predicted reserves: 94.96504596800489


The models have been successfully evaluated. From our evaluation, the average volume of predicted reserves are 92.6, 68.7 and 95.0 for the 1st, 2nd and 3rd regions respectively. The RMSE shows the degree of variation the prediction is away from the correct results. The model is most accurate in the 2nd region as we can see the lowest RMSE

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Let's fix previous comments and get back to this conclusion one more time.
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment V2</b> <a class="tocSkip"></a>

I think the model is most accurate in the 2nd region
</div>


## Preparation for profit calculation

### Create sample of 500 points for each region

#### 1st region

In [41]:
# Create sample
s1 = d1_valid.reset_index(drop=True)

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

- For our future tasks we need to use only validation data.
- Also, We can use parameter n instead of frac.
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment V2</b> <a class="tocSkip"></a>

Good, thank you
</div>

#### 2nd region

In [42]:
# Create sample
s2 = d2_valid.reset_index(drop=True)

#### 3rd region

In [43]:
# Create sample 
s3 = d3_valid.reset_index(drop=True)

### Store all key values for calculations in separate variables.

In [44]:
# Store all key values for calculations in separate variables.

dev_budget = 100000000 # The budget for development of 200 oil wells

cost_of_one = dev_budget/200

one_barr_rev = 4.5 # One barrel of raw materials brings 4.5 USD of revenue 

### Calculate the volume of reserves sufficient for developing a new well without losses

In [45]:
# Calculate volume sufficient for new well
vol_for_well = cost_of_one/one_barr_rev
vol_for_well = vol_for_well/1000
print('the volume of reserves sufficient for developing a new well without losses', vol_for_well)

the volume of reserves sufficient for developing a new well without losses 111.11111111111111


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

OK, good
</div>

### Average volume of reserves in each region

#### 1st region

In [46]:
# Ave vol of reserves
ave_vol_res_1 = data_1['product'].mean()
print('Ave vol:',ave_vol_res_1 )

Ave vol: 92.50000000000001


#### 2nd region

In [47]:
# Ave vol of reserves
ave_vol_res_2 = data_2['product'].mean()
print('Ave vol:',ave_vol_res_2 )

Ave vol: 68.82500000000002


#### 3rd region

In [48]:
# Ave vol of reserves
ave_vol_res_3 = data_3['product'].mean()
print('Ave vol:',ave_vol_res_3 )

Ave vol: 95.00000000000004


The variables are ready to be used for profit calculation. From our above calculations, a minimum volume of 111 thousand should be available at each location to avoid running at a loss when creating new oil wells. The average volume of reserves in the 3 regions are lower than this.

## Calculate profit from a set of selected oil wells and model predictions

### Get predictions for each sample

#### S1

In [49]:
# Get sample features
s1_features = s1.drop(['product', 'id'], axis=1)

In [50]:
# Get predictions
s1_pred = model_1.predict(s1_features)

#### S2

In [51]:
# Get sample features
s2_features = s2.drop(['product', 'id'], axis=1)

In [52]:
# Get predictions
s2_pred = model_2.predict(s2_features)

#### S3

In [53]:
# Get sample features
s3_features = s3.drop(['product', 'id'], axis=1)

In [54]:
# Get predictions
s3_pred = model_3.predict(s3_features)

### Create function to calculate profit

In [55]:
# Create profit function
def profit(df, predictions, revenue_per_barrel):
    prof_list = []
    ovr_prof = 0
    
    high_pred = pd.Series(predictions)
    high_pred = high_pred[df.index].sort_values(ascending=False).head(200)
    df_new_prod = df['product']
    target = df_new_prod[high_pred.index]
    
    for value in target:
        prof = (value * revenue_per_barrel * 1000) - cost_of_one 
        prof_list.append(prof)
    ovr_prof = pd.Series(prof_list).sum()
    return ovr_prof

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

There are several issues:
- We should calculate profit as revenue - costs
- To get revenue we need to: "Summarize the target volume of reserves in accordance with these predictions". So, select the wells with the highest values of predictions but use target volume of reserves to calculate revenue.
- Could you comment why we have this magic number 1000.

</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>

We are multiplying by 1000 because volume of reserves is in thousand barrels.
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment V2</b> <a class="tocSkip"></a>

Good, but I'd suggest to put the process of selecting wells inside of the function. It could help us further.
</div>

### Calculate profit for each region

#### s1_high

In [56]:
print('Profit for this region:', profit(s1, s1_pred, 4.5))

Profit for this region: 26385004.701609157


#### s2_high

In [57]:
print('Profit for this region:', profit(s2, s2_pred, 4.5))

Profit for this region: 24150866.96681509


#### s3_high

In [58]:
print('Profit for this region:', profit(s3, s3_pred, 4.5))

Profit for this region: 27138232.94735464


From the above calculations, the best region for oil well development is the 2nd region. This region brings in the most profit of all 3 regions at a whooping 7.7 million USD.

## Calculate risks and profit for each region

### Create bootstrapping function

In [59]:
# Create bootstrapping function
def boot_func(df, prediction, revenue_per_barrel):
    
    state = np.random.RandomState(12345)
    
    prof_distr = []
    
    loss = []
    
    for i in range(1000):
        boot_samp = df.sample(n=500, replace=True, random_state= state)
        rev = profit(boot_samp, prediction, revenue_per_barrel)
        prof_distr.append(rev)
    
    
    prof_distr = pd.Series(prof_distr)
    
    
    lower = prof_distr.quantile(0.025)
    upper = prof_distr.quantile(0.975)
    
    for value in prof_distr:
        if value - vol_for_well < 0:
            loss.append(value)
    
    loss = pd.Series(loss)
    
    print(f'Risk of losses: {(prof_distr < 0).mean():.2%}')
    
    
    print('Average profit for this region:', prof_distr.mean())
    
    print(f'95% confidence interval:{lower} - {upper}')
    
    return 'Analysis complete'

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

- The same issue with the way to calculate revenue. We need to pass not predictions, but target values selected by predictions.
- To calculate Risk of losses, we need to find a fraction of values in prof_distr that is lower than 0. 
- var = print('something') is not a correct form. Print can't return anything. That's why we see (None, None, None) bellow. We can just use print and don't use return in this function.
</div>

<div class="alert alert-block alert-danger">
<b>Reviewer's comment V2</b> <a class="tocSkip"></a>

- The idea behind the bootstrap is the following. We should: randomly select 500 wells, select best 200 based by our predicion, calculate profit and repeat the process 1000 times. Now you pass best 200 every time and select 25% of them to calculate the profit which is not correct.
- To calculate Risk of losses, we need to use this: (prof_distr < 0).mean(). And add 2 fraction numbers (:.0% -> :.2%), please.

</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment V3</b> <a class="tocSkip"></a>

Good, thank you
</div>


### Evaluation of each region

#### 1st Region

In [60]:
# Evaluate region with bootstrapping function
print(boot_func(s1, s1_pred,  4.5))

Risk of losses: 13.50%
Average profit for this region: 3136758.1596374847
95% confidence interval:-2127877.66605007 - 8352807.233802227
Analysis complete


#### 2nd Region

In [61]:
# Evaluate region with bootstrapping function
print(boot_func(s2, s2_pred, 4.5))

Risk of losses: 1.70%
Average profit for this region: 4651514.195956517
95% confidence interval:320367.42850019847 - 8643762.015334848
Analysis complete


#### 3rd Region

In [62]:
# Evaluate region with bootstrapping function
print(boot_func(s3, s3_pred, 4.5))

Risk of losses: 8.10%
Average profit for this region: 4143469.2794738654
95% confidence interval:-1490665.2762989856 - 9723391.679202836
Analysis complete


From the above evaluation, the 2nd region has the lowest risk of losses and the highest average profit and so this is the best for oil well development.

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Let's fix previous comments and get back to this conclusion one more time.
</div>

<div class="alert alert-block alert-success">

<b>Reviewer's general comment</b> <a class="tocSkip"></a>

You've done a great job! Let's fix some issues and the project will be accepted.
</div>



<div class="alert alert-block alert-danger">

<b>Reviewer's general comment</b> <a class="tocSkip"></a>

What we should pay attention to:

- use transform method on validation features
- check which model you apply in 3.5.2
- for profit calculation use only validation data
- fix profit() function
- fix boot_func()
- review comments


</div>

<div class="alert alert-block alert-danger">
<b>Reviewer's general comment V2</b> <a class="tocSkip"></a>

Thank you for taking the time to improve the project. Let's fix the final step and we should be fine.

</div>

<div class="alert alert-block alert-success">
<b>Reviewer's general comment V3</b> <a class="tocSkip"></a>

Good job. Now the project is accepted. You can move on to the next module. Congratulations!
</div>